File: backend/app/core/geo.py
"""أدوات جغرافية: تحويل الإحداثيات إلى/من نقاط PostGIS وحساب المسافة"""
from math import asin, cos, radians, sin, sqrt

from geoalchemy2.elements import WKTElement
from geoalchemy2.shape import to_shape

SRID = 4326  # نظام الإحداثيات WGS84 (خط الطول/العرض)


def make_point(lat: float, lng: float) -> WKTElement:
    """ينشئ نقطة جغرافية صالحة للتخزين في عمود Geography.

    ملاحظة: ترتيب WKT هو POINT(lng lat) — خط الطول أولاً.
    """
    return WKTElement(f"POINT({lng} {lat})", srid=SRID)


def read_point(value: object) -> tuple[float, float] | None:
    """يقرأ نقطة مخزّنة ويُعيد (lat, lng)، أو None إن لم تكن موجودة."""
    if value is None:
        return None
    shape = to_shape(value)  # type: ignore[arg-type]
    return shape.y, shape.x


def haversine_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """المسافة التقريبية بين نقطتين بالكيلومترات (صيغة هافرسين)."""
    r = 6371.0  # نصف قطر الأرض بالكيلومتر
    d_lat = radians(lat2 - lat1)
    d_lng = radians(lng2 - lng1)
    a = sin(d_lat / 2) ** 2 + cos(radians(lat1)) * cos(radians(lat2)) * sin(d_lng / 2) ** 2
    return round(2 * r * asin(sqrt(a)), 2)

File: mobile/src/utils/geo.ts
/** حسابات جغرافية بسيطة على العميل (تقدير المسافة/الوقت لتتبّع السائق) */
import type { Location } from "../api/types";

const R_KM = 6371;
const toRad = (d: number) => (d * Math.PI) / 180;

/** المسافة بالكيلومترات بين نقطتين (صيغة haversine) */
export function haversineKm(a: Location, b: Location): number {
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const lat1 = toRad(a.lat);
  const lat2 = toRad(b.lat);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.sin(dLng / 2) ** 2 * Math.cos(lat1) * Math.cos(lat2);
  return 2 * R_KM * Math.asin(Math.min(1, Math.sqrt(h)));
}

/** وقت وصول تقديري بالدقائق بافتراض سرعة حضرية وسطية للدرّاجة (~22 كم/س) */
export function etaMinutes(distanceKm: number, speedKmh = 22): number {
  return Math.max(1, Math.round((distanceKm / speedKmh) * 60));
}
